pennylane_qiskit/converter.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+QuanctumCircuit converter module
+================================
+
+.. currentmodule:: pennylane_qiskit.load
+
+This module contains functions for converting Qiskit QuantumCirducit objects
+PennyLane templates.
+"""
+import warnings
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter
+from qiskit.exceptions import QiskitError
+
+import pennylane.ops.qubit as pennylane_ops
+from pennylane_qiskit.qiskit_device import QISKIT_OPERATION_MAP
+
+# pylint: disable=too-many-instance-attributes
+
+inv_map = {v.__name__: k for k, v in QISKIT_OPERATION_MAP.items()}
+
+
+def check_parameter_bound(param):
+    """Utility function determining if a certain parameter in a QuantumCircuit has
+    been bound.
+
+    Args:
+        param: the parameter to be checked
+
+    Returns:
+        param: the parameter after the check
+    """
+    if isinstance(param, Parameter):
+        raise ValueError("The parameter {} was not bound correctly.".format(param))
+    return param
+
+
+def load(quantum_circuit: QuantumCircuit):
+    """Returns a PennyLane template created based on the input QuantumCircuit or QASM string.
+    Warnings are created for each of the QuantumCircuit instructions that were
+    not incorporated in the PennyLane template.
+
+    Args:
+        quantum_circuit (qiskit.QuantumCircuit): the QuantumCircuit to be converted
+
+    Returns:
+        function: the new PennyLane template
+    """
+
+    def _function(params: dict = None, wire_shift: int = 0):
+        """Returns a PennyLane template created based on the input QuantumCircuit.
+        Warnings are created for each of the QuantumCircuit instructions that were
+        not incorporated in the PennyLane template.
+
+        Args:
+            params (dict): specifies the parameters that need to bound in the
+            QuantumCircuit
+
+            wire_shift (int): the shift to be made with respect to the wires
+            already specifiec in the QuantumCircuit
+            e.g. if the QuantumCircuit acts on wires [0, 1] and if wire_shift == 1,
+            then the returned Pennylane template will act on wires [1, 2]
```

Comment:
Thinking about this more, this should be relatively easy to port it to a `wires=` like approach.

For example, if it's a 3 qubit quantum circuit, and the user converts it into a template with `wires=[2,0,4]`, then:
```python
qc_wires = [q.index for q in quantum_circuit.qubits]
wire_map = {k:v for k, v in zip(qc_wires, wires)}
```

and then, further down below it becomes
```python
operation(wires=[wire_map[w] for w in wires])
```

Is this a small enough change to make here now?

Reply:
Thanks for that @josh146! I pushed a modification that includes this as well with new tests.

This made me realize, that ``qubit.index`` might not uniquely identify the wires operations act on. The following is an example for this:
```python
qr1 = QuantumRegister(2)
qr2 = QuantumRegister(1)
qc = QuantumCircuit(qr1, qr2)
wires = [q.index for q in circuit.qubits]
```
In this example, ``wires == [0, 1, 0, 1]``, therefore the ``(qubit.register.name, qubit.index)`` tuple is used to distinguish between wires.

Modified the code according to this as well.